blueprints/flexible-topologies.md

```diff
@@ -0,0 +1,395 @@
+
+### Security Impact
+
+Probably the biggest impact to security will be the custom Topologies themselves, specifically in terms of breaking the assumptions of the existing 2-tier CDN architecture. There will no longer be a single, global parent hierarchy, so things like firewalls/ACLs that allow caches to communicate with each other may need to be updated to account for custom parent hierarchies.
+
+Creating or updating Topologies should be restricted to users with the `operations` role or higher, and Topologies are CDN-wide -- they do not belong to a particular tenant.
+
+### Upgrade Impact
+
+This feature will require a database migration to create new tables and add a new field to the `deliveryservice` table, but existing data does not need to be modified or migrated. Therefore, rolling back the database migration will not cause any data loss until Topologies are actually created and assigned. Additionally, since this feature will not remove any existing tables or columns, the new database schema should be backwards-compatible with the previous version of Traffic Ops. However, this blueprint cannot make any claim as to the backwards-compatibility of schema changes required by *other* features, which would affect the overall backwards-compatibility of the entire release.
+
+This feature will not require components to be upgraded in a specific order, and no special manual steps will be required before, during, or after the upgrade is done.
+
+### Operations Impact
+
+One of the bigger day-to-day operational impacts this feature will have is in the assignment of Delivery Services to Topologies instead of to individual `EDGE` caches. In theory, if the number of unique Topologies is kept as low as possible, it should be easier to choose a specific Topology for a Delivery Service than to assign it to individual caches (unless the default approach is to just assign delivery services to *all* caches). This could be made easier by defining a kind of *default* Topology which is assigned to until it is determined that the default Topology does not meet the Delivery Service's requirements.
+
+Until legacy `deliveryservice_server` assignments are fully removed in favor of Topology-based assignments, there will be extra operational overhead due to having two different ways to assign Delivery Services, but this overhead should be temporary as CDN operators should migrate all Delivery Services to Topologies as soon as possible. That one-time manual migration should also be considered an operational impact and should be scheduled to take place sometime after the upgrade is completed.
```

Comment:
this default topology thing is kind of interesting and maybe we can pursue in future phases of this feature? 
i.e.

cdn1, default topology=foo
cdn2, default topology=bar

this isn't something that can be seeded with TO as a topology consists of cgs and servers that are company specific but maybe giving operations or admin users the ability to select a default topology for each cdn would be very beneficial imo.

Reply:
and then by calling something like GET /topologies/default?type=http_live it would return the default, if any, which could be selected in a new DS form in TP.